Enable Deepspeed Ulysses for Wan

mengker33 opened the pull request.

mengker33 committed:
Wan: Enable deepspeed ulysses for ti2v pipeline

mengker33 committed:
Deepspeed: Fix uneven head sequence parallelism bug

This bug has been fixed in upstream deepspeed but not included in habana
deepspeed 1.22. See upstream fix commit:
https://github.com/deepspeedai/DeepSpeed/commit/ecb4bf32614fa8d6d4433ea579ca867008d34140

mengker33 committed:
README: Add Wan i2v example using deepspeed ulysses

ikurtchen commented on DeepSpeed/deepspeed/sequence/layer.py:
```diff
@@ -184,7 +184,7 @@ def single_all_to_all(input, scatter_idx, gather_idx, batch_dim_idx, group, asyn
     num_heads = input.shape[2]
 
-    if get_num_kv_heads() is not None or num_heads % seq_world_size != 0:
+    if get_num_kv_heads() is not None or (num_heads % seq_world_size != 0 and not scatter_idx < 2):
```
What's the case for adding this check? Will uneven heads use even all2all in this case?

mengker33 replied:
For last output all2all, head is split and seq is complete, in this case there is no need to check num_heads % seq_world_size. See https://github.com/HabanaAI/optimum-habana-fork/blob/aice/v1.22.0/DeepSpeed/deepspeed/sequence/layer.py#L405.

ikurtchen commented on examples/stable-diffusion/image_to_video_generation.py:
```diff
@@ -295,6 +304,14 @@ def main():
     if args.bf16:
         kwargs["torch_dtype"] = torch.bfloat16
 
+    if args.context_parallel_size > 1 and parallel_state.is_unitialized():
```
Is this a typo? "is_unitialized", is_uninitialized?

mengker33 replied:
It is a typo from oh itself and I don't prefer to correct it here. See https://github.com/HabanaAI/optimum-habana-fork/blob/aice/v1.22.0/optimum/habana/distributed/parallel_state.py#L353

ikurtchen commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -206,8 +208,94 @@ def __init__(self, fusedSDPA):
+
+class GaudiDistributedAttention(torch.nn.Module):
+    def __init__(
+        self, hpu_module_fsdpa: ModuleFusedSDPA, scale, attention_dropout, enable_recompute, flash_attention_fp8
```
scale, attention_dropout, enable_recompute, flash_attention_fp8 are not used?

mengker33 replied:
You're right, let me remove these unused parameters.

ikurtchen commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -565,14 +668,27 @@ def _native_attention(
-        out = out.permute(0, 2, 1, 3)
+
+        if self.fused_scaled_dot_product_attention_distributed:
+            out = self.fused_scaled_dot_product_attention_distributed(
```
Why the FusedSDPA call is different from not SP case?

mengker33 replied:
make them the aligned call.

ikurtchen commented on optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py:
```diff
@@ -176,6 +177,22 @@ def prepare_latents(
 
+        def adjust_num_frame(t, h, w, cp_size):
+            # It is to make sure seq_len can be devided by cp_size.
+            MAX_T = t + 100
```
Is it possible not hard code 100 to find the padding size?

mengker33 replied:
Move the padding after patching.

ikurtchen commented on optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py:
```diff
@@ -176,6 +177,22 @@ def prepare_latents(
+        p_t, p_h, p_w = self.transformer.config.patch_size
+        tmp_seq_len = (num_latent_frames // p_t) * (latent_height // p_h) * (latent_width // p_w)
+        if tmp_seq_len % cp_size != 0:
+            num_latent_frames = adjust_num_frame(
```
Dose it need to remember how many padding added here, so as to mask attn and remove padding later?

mengker33 replied:
I have added the attention_mask after padding. But considering padding size is very small compared to original seq_len, mask will not be applied by default.

Wei-Lin-Intel commented on optimum/habana/diffusers/models/wan_transformer_3d.py:
```diff
@@ -66,6 +68,25 @@ def WanTransformer3DModleForwardGaudi(
+        start = sp_seq_len * parallel_state.get_sequence_parallel_rank()
+        end = sp_seq_len * (parallel_state.get_sequence_parallel_rank() + 1)
+
+        hidden_states = hidden_states[:, start:end, :]
```
Padded sequence length in SDPA without a mask would trigger the accuracy issue. It is necessary to add mask here if `seq_len` is padded. Use bool tensor for the mask can save some memory.

mengker33 committed:
Enable deepspeed ulysses for Wan t2v pipeline

mengker33 committed:
Add attn_mask when there is padding in cp case

When seq_len cannot be divided by cp_size, there is padding in seq_len.
and attn_mask is helpful to reduce the padding impact on accuracy.
Considering padding is usually very small compared to seq_len in the
diffuser case, we don't apply the attn_mask by default.
The attn_mask can be activated by setting env var CP_USE_MASK.

mengker33 committed:
README: Add Wan2.2 t2v example

yingjie-han commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -206,8 +208,92 @@ def __init__(self, fusedSDPA):
+        valid_sequence_lengths,
+        padding_side="left",
+    ):
+        query, key, value = (x.permute(0, 2, 1, 3) for x in (query, key, value))
```
单卡跑的时候，这里比原始代码多了permute。fused_scaled_dot_product_attention输入的时候先transpose,这里在permute，这样是否会带来不必要的开销。

Wei-Lin-Intel replied:
Add a `.contiguous()` after permute may avoid DMA cost in SDPA

mengker33 replied:
It is moving permute from ModuleSDPA outside to inside, the same for single-card and multi-card.

Wei-Lin-Intel commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -262,17 +348,20 @@ def __call__(
-            is_casual=False,
-            scale=None,
-            softmax_mode=softmax_mode,
+            query.transpose(1, 2),
```
same here, add `.contiguous()` here

huijuanzh commented:
LGTM, if update some performance data compared with 1/2/4 card would be good.

yingjie-han commented on optimum/habana/diffusers/models/wan_transformer_3d.py:
```diff
@@ -66,6 +70,47 @@ def WanTransformer3DModleForwardGaudi(
+        start = sp_seq_len * parallel_state.get_sequence_parallel_rank()
+        end = sp_seq_len * (parallel_state.get_sequence_parallel_rank() + 1)
+
+        hidden_states = hidden_states[:, start:end, :]
```
Here attention_mask is not splitted. I think attention_mask also need to be splited by sp , what do you think?

mengker33 replied:
attention_mask is only used in deepspeed ulysses self-attention case where in ModuleFusedSDPA, q/k/v are already gathered along seq and split along head, so it is ok to create the attention_mask  with full seq len here.

mengker33 commented:
> LGTM, if update some performance data compared with 1/2/4 card would be good.

Linwei suggested to use classic SP to reduce memory consumption. I will include SP in this PR too and provide perf data for both deepspeed ulysses and SP soon.

mengker33 committed:
Wan: Add traditional SP in wan attention

Co-authored-by: Wei Lin <wei2.lin@intel.com>

mengker33 committed:
Wan: Replace vae WanBlockAttention with FusedSDPA

mengker33 committed:
README: Update readme after adding SP support

mengker33 commented:
<img width="1518" height="88" alt="image" src="https://github.com/user-attachments/assets/47f1f12f-3560-468e-bbe7-429bf269f8a2" />
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>
<!--StartFragment-->

<div style='direction:ltr'>


num_card | SP:   generated_time | DS:   generated_time
-- | -- | --
1 | 245s | 245s
2 | Avg:   145s | Avg:141s
4 | Avg:74.75s | Avg: 71s
8 | Avg: 59s | Avg: 55s



</div>

<!--EndFragment-->
</body>

</html>

@Wei-Lin-Intel @huijuanzh Please help to review again, thank you:)

Wei-Lin-Intel reviewed (approved):
LGTM

huijuanzh commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -262,17 +348,20 @@ def __call__(
-            is_casual=False,
-            scale=None,
-            softmax_mode=softmax_mode,
+            query.transpose(1, 2).contiguous(),
```
why CogVideoXAttnProcessorGaudi need to change? Since you put qkv transpose in ModuleFusedSDPA, need to check other function which not use sp or ulysses.

huijuanzh commented:
LGTM
